Disallow trailing commas when collection type sugar is parsed as exprs.

allevato opened the pull request:
When the type checker simplifies an `ArrayExpr` or `DictionaryExpr` to the
equivalent `TypeRepr`, it doesn't check whether a trailing comma is present.
Since a trailing comma is allowed for normal expressions, this permits a
strange syntactic inconsistency:

```swift
let x = [Int,]()  // ok
let x: [Int,]     // error

let y = [String: Int,]()  // ok
let y: [String: Int,]     // error
```

This change makes it consistently an error when collection type sugar
parsed as an expression has a trailing comma.

allevato commented:
This is a minimal change to make the trailing comma an error, but the resulting diagnostics aren't great (see the tests). Rather than returning `nullptr` here, it would probably make more sense to diagnose the trailing comma and provide a fix-it, and then let the type checker continue. Thoughts?

futurejones commented:
Can one of the admins verify this patch?

CodaFi commented:
Sure, the diagnostics aren't fantastic, but I think still worth taking this as a matter of correctness.

allevato commented:
I'm happy to either mark this PR as ready for review now and then follow it up with a second PR with improved diagnostics, or I can do everything in this PR—whatever you prefer. I'm not in a rush to merge this and I'd rather do the right/clean thing.

allevato commented:
I finally had some time to dust this off again, so I went ahead and improved the diagnostic quality. Trailing commas in array/dictionary type sugar diagnosed in TypeCheckConstraints.cpp now emit the same diagnostics that the parser emits when it handles the same situation in type contexts.

For example:

```
<stdin>:4:21: error: expected ']' in dictionary type
let y = [String: Int,]()
                    ^
<stdin>:4:9: note: to match this opening '['
let y = [String: Int,]()
        ^
```

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

allevato committed:
Disallow trailing commas when collection type sugar is parsed as exprs.

When the type checker simplifies an `ArrayExpr` or `DictionaryExpr` to the
equivalent `TypeRepr`, it doesn't check whether a trailing comma is present.
Since a trailing comma is allowed for normal expressions, this permits a
strange syntactic inconsistency:

```swift
let x = [Int,]()  // ok
let x: [Int,]     // error

let y = [String: Int,]()  // ok
let y: [String: Int,]     // error
```

This change makes is consistently an error when collection type sugar
parsed as an expression has a trailing comma.

allevato commented:
@swift-ci please test

allevato commented:
I've rebased this on top of the `main` branch.

@CodaFi , if you think it's still good to have this for consistency/correctness, can you review it? Thanks!

slavapestov commented:
What if instead of diagnosing these expressions we left them as-is without converting to type sugar?
